Parse Large Numbers and Scientific Notation

s9latimm opened the pull request:
Allow numbers to be written in scientific notation in XML / constraints and handle out-of-range integers.

related se-sic/VaRA/issues/813

closes se-sic/VaRA/issues/814
closes se-sic/VaRA/issues/815

s9latimm committed:
Handle scientific notation and unsigned.

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/ConstraintParser.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModelParser.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/ConstraintParser.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/FeatureTreeNode.cpp
M	unittests/Feature/NumericFeature.cpp
M	unittests/Feature/Relationship.cpp
M	unittests/resources/CMakeLists.txt
A	unittests/resources/xml/test_numbers.xml

s9latimm committed:
Tidy.

M	unittests/Feature/FeatureModelParser.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Merge remote-tracking branch 'origin/vara-dev' into f-parse-number

M	.clang-tidy
M	.github/workflows/build.yml
M	.gitignore
M	CMakeLists.txt
M	bindings/python/tests/feature_model/test_feature_model.py
M	bindings/python/tests/writer/test_xml_writer.py
M	bindings/python/vara-feature/CMakeLists.txt
M	bindings/python/vara-feature/pybind_FeatureModel.cpp
D	bindings/python/vara-feature/pybind_FeatureModelXmlParser.cpp
M	bindings/python/vara-feature/pybind_Init.cpp
M	bindings/python/vara-feature/pybind_common.h
M	docs/images/constraints.png
M	external/googletest
M	external/pybind11
A	include/vara/Configuration/Configuration.h
M	include/vara/Feature/ConstraintParser.h
A	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
M	include/vara/Feature/FeatureSourceRange.h
M	include/vara/Feature/FeatureTreeNode.h
A	include/vara/Utils/Result.h
M	include/vara/Utils/VariantUtil.h
M	lib/CMakeLists.txt
A	lib/Configuration/CMakeLists.txt
A	lib/Configuration/Configuration.cpp
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelTransaction.cpp
A	scripts/CMakeLists.txt
A	scripts/ci_utils/CMakeLists.txt
A	scripts/ci_utils/prepare_coverage_data.py
M	tools/fm-viewer/FeatureModelViewer.cpp
M	unittests/CMakeLists.txt
A	unittests/Configuration/CMakeLists.txt
A	unittests/Configuration/Configuration.cpp
A	unittests/Configuration/ConfigurationOption.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelSxfm.cpp
M	unittests/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/NumericFeature.cpp
M	unittests/Utils/CMakeLists.txt
A	unittests/Utils/Result.cpp
M	unittests/Utils/VariantUtil.cpp
M	unittests/resources/CMakeLists.txt
M	unittests/resources/sxfm/test.sxfm
A	unittests/resources/sxfm/test_wrong_constraint_format.sxfm

s9latimm committed:
Fix bindings.

M	bindings/python/vara-feature/pybind_Feature.cpp
M	bindings/python/vara-feature/pybind_FeatureModel.cpp

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -570,10 +570,10 @@ class PrimaryConstraint : public Constraint {
 class PrimaryIntegerConstraint : public PrimaryConstraint {
 public:
-  PrimaryIntegerConstraint(int Value)
+  PrimaryIntegerConstraint(long Value)
```
```suggestion
  PrimaryIntegerConstraint(int64_t Value)
```

I would use `int64_t` instead
a) it makes the size clear
b) it also works for 32bit latforms

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -230,7 +230,10 @@ class ConstraintLexer {
   static ResultTy munchNumber(const llvm::StringRef &Str) {
-    auto Munch = Str.take_while([](auto C) { return llvm::isDigit(C); });
+    auto Munch = Str.take_while([](auto C) {
+      return llvm::isDigit(C) || C == 'e' || C == '+' || C == '-';
```
did you run `toLower` before? If not, 'E' could also happen

`12E3`

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -497,6 +499,30 @@ class ConstraintParser {
   }
 
+  static std::unique_ptr<Constraint> parseUnsigned(llvm::StringRef Str) {
+    if (Str.contains_lower('e')) {
```
Did they rename the method? In the StringRef API I only find `contains_insensitive` and not `contains_lower`

s9latimm replied:
This method was only renamed six months ago (https://github.com/llvm/llvm-project/commit/3eed57e7ef7da5eda765ccc19fd26fb8dfcd8d41), therefore i use the old name with our LLVM version.
This will probably break at some point, as we update the framework.

vulder replied:
:+1: I thought something like this :)

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -497,6 +499,30 @@ class ConstraintParser {
+      return std::make_unique<PrimaryIntegerConstraint>(round(Number));
+    }
+
+    try {
+      return std::make_unique<PrimaryIntegerConstraint>(std::stol(Str.str()));
+    } catch (std::invalid_argument &_) {
+      llvm::errs() << "Failed to parse number '" << Str << "'\n";
+      return nullptr;
+    } catch (std::out_of_range &_) {
+      llvm::errs() << "Number too large.\n";
+      return std::make_unique<PrimaryIntegerConstraint>(
+          std::numeric_limits<long>::max());
+    }
```
Do we have a way to implement this without exception handling? LLVM builds without exceptions and vara-feature is embedded into llvm/vara

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -123,6 +123,32 @@ FeatureModelXmlParser::parseConfigurationOption(xmlNode *Node,
   return Ok();
 }
 
+long FeatureModelXmlParser::parseNumber(llvm::StringRef Str) {
+  if (Str.contains_lower('e')) {
+    double Number;
+    std::stringstream Stream(Str.str());
+    Stream >> Number;
+    if (Stream.fail()) {
+      llvm::errs() << "Failed to parse number '" << Str << "'\n";
+      return 0;
+    }
+    return (long)round(Number);
+  }
+
+  try {
+    return std::stol(Str.str());
+  } catch (std::invalid_argument &_) {
+    llvm::errs() << "Failed to parse number '" << Str << "'\n";
+    return 0;
+  } catch (std::out_of_range &_) {
+    llvm::errs() << "Number out of range.\n";
+    if (Str.startswith("-")) {
+      return std::numeric_limits<long>::min();
+    }
+    return std::numeric_limits<long>::max();
+  }
+}
```
This seems to be a duplicate of ConstrantParser `parseUnsigned`, we should refactor this into a single impl

s9latimm committed:
Without exceptions.

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/ConstraintParser.h
M	include/vara/Feature/Feature.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/ConstraintParser.cpp

s9latimm committed:
Unused import.

M	include/vara/Feature/ConstraintParser.h

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -257,6 +260,31 @@ class ConstraintLexer {
+    // If we encounter scientific notation we try to parse the number as double.
+    double Double = 0;
+    if (!Str.getAsDouble(Double)) {
+      return round(Double);
```
We need to add a check here, to war should a user pass an value that's to large.

s9latimm committed:
Fix double overflow.

M	include/vara/Feature/ConstraintParser.h
M	unittests/Feature/ConstraintParser.cpp

s9latimm committed:
Add error location.

M	include/vara/Feature/ConstraintParser.h
M	lib/Feature/FeatureModelParser.cpp
